File: clients/java/src/main/java/io/keysync/KeySyncException.java
package io.keysync;

/**
 * Exception thrown by KeySync operations.
 *
 * <p>Contains an error code ({@link KeySyncError}) and a human-readable message.
 *
 * <ul>
 *   <li>{@code NOT_FOUND} — the secret does not exist in any scope</li>
 *   <li>{@code KEYCHAIN_ERROR} — the native keychain tool failed</li>
 *   <li>{@code UNSUPPORTED_PLATFORM} — the current OS is not supported</li>
 * </ul>
 */
public class KeySyncException extends RuntimeException {

    private final KeySyncError error;

    /**
     * Creates a new KeySyncException.
     *
     * @param error   the error type
     * @param message a human-readable description
     */
    public KeySyncException(KeySyncError error, String message) {
        super(message);
        this.error = error;
    }

    /**
     * Creates a new KeySyncException with a cause.
     *
     * @param error   the error type
     * @param message a human-readable description
     * @param cause   the underlying cause
     */
    public KeySyncException(KeySyncError error, String message, Throwable cause) {
        super(message, cause);
        this.error = error;
    }

    /** Returns the error type. */
    public KeySyncError getError() {
        return error;
    }

    /** Returns the error code string (e.g. "notFound"). */
    public String getErrorCode() {
        return error.getCode();
    }
}

File: clients/csharp/KeySyncException.cs
namespace KeySync;

/// <summary>
/// Exception thrown by the KeySync client library.
/// </summary>
public class KeySyncException : Exception
{
    /// <summary>
    /// The error code identifying the type of failure.
    /// </summary>
    public KeySyncError ErrorCode { get; }

    internal KeySyncException(KeySyncError errorCode, string message)
        : base(message)
    {
        ErrorCode = errorCode;
    }

    internal KeySyncException(KeySyncError errorCode, string message, Exception innerException)
        : base(message, innerException)
    {
        ErrorCode = errorCode;
    }

    /// <summary>
    /// Creates a <see cref="KeySyncException"/> for "secret not found".
    /// </summary>
    internal static KeySyncException NotFound()
        => new(KeySyncError.NotFound, "secret not found");

    /// <summary>
    /// Creates a <see cref="KeySyncException"/> for a keychain-level error.
    /// </summary>
    internal static KeySyncException KeychainError(string detail)
        => new(KeySyncError.KeychainError, $"keychain error: {detail}");

    /// <summary>
    /// Creates a <see cref="KeySyncException"/> for a keychain-level error
    /// wrapping an inner exception.
    /// </summary>
    internal static KeySyncException KeychainError(string detail, Exception innerException)
        => new(KeySyncError.KeychainError, $"keychain error: {detail}", innerException);

    /// <summary>
    /// Creates a <see cref="KeySyncException"/> for unsupported platforms.
    /// </summary>
    internal static KeySyncException UnsupportedPlatform()
        => new(KeySyncError.UnsupportedPlatform, "keychain access not available on this platform");
}
